apis/python/src/tiledbsoma/_dataframe.py

```diff
@@ -412,9 +412,22 @@ def write(
             n = len(col)
             cols_map = dim_cols_map if name in dim_names_set else attr_cols_map
             if pa.types.is_dictionary(col.type) and col.num_chunks != 0:
-                cols_map[name] = col.chunk(0).indices.to_pandas()
+                attr = self._handle.schema.attr(name)
+                if attr.enum_label is not None:
+                    # Normal case: writing categorical data to categorical schema.
+                    cols_map[name] = col.chunk(0).indices.to_pandas()
+                else:
+                    # Schema is non-categorical but the user is writing categorical.
+                    # Simply decategoricalize for them.
+                    pvalues = []
+                    for chunk in col.chunks:
+                        pvalues += [
+                            chunk.dictionary[i.as_py()].as_py() for i in chunk.indices
```

Comment:
Apologies if I'm misreading this, but this appears to be iterating over _every_ element in the column in a Python loop.  That is going to be unbelievably slow and use a great deal of memory for anything other than very small Arrow arrays.

There are built-ins for doing this.  I believe you want (please confirm with tests) to replace the entire thing with:

> cols_map[name] = pa.chunked_array([chunk.dictionary_decode() for chunk in col.chunks])